src/translator/translation_model.h

```diff
@@ -0,0 +1,122 @@
+#ifndef SRC_BERGAMOT_TRANSLATION_MODEL_H_
+#define SRC_BERGAMOT_TRANSLATION_MODEL_H_
+
+#include <string>
+#include <vector>
+
+#include "batch.h"
+#include "batching_pool.h"
+#include "common/utils.h"
+#include "data/shortlist.h"
+#include "definitions.h"
+#include "parser.h"
+#include "request.h"
+#include "text_processor.h"
+#include "translator/history.h"
+#include "translator/scorers.h"
+#include "vocabs.h"
+
+namespace marian {
+namespace bergamot {
+
+/// A TranslationModel is associated with the translation of a single language direction. Holds the graph and other
+/// structures required to run the forward pass of the neural network, along with preprocessing logic (TextProcessor)
+/// and a BatchingPool to create batches that are to be used in conjuction with an instance.
+///
+/// Assuming parameters are provided correctly, the following is an example of use of API provided by this class to
+/// complete translation:
+///
+/// ```cpp
+/// TranslationModel model(...);
+/// auto request = model.makeRequest(...);
+/// model.enqueueRequest(request);
+/// Batch batch;
+/// while ( generateBatch(batch)) {
+///      model.translateBatch(batch);
+/// }
+/// ```
+///
+/// Thread-safety is not handled here, but the methods are available at granularity enough to be used in threaded async
+/// workflow for translation.
+
+class TranslationModel {
+ public:
+  using Config = Ptr<Options>;
+
+  /// Equivalent to options based constructor, where `options` is parsed from string configuration. Configuration can be
+  /// JSON or YAML. Keys expected correspond to those of `marian-decoder`, available at
+  /// https://marian-nmt.github.io/docs/cmd/marian-decoder/
+  TranslationModel(const std::string& config, MemoryBundle&& memory, size_t replicas = 1)
```

Comment:
This replicas thing is ugly.  I get why you are doing it this way since we don't have shared workspace memory yet.